File: Prototype/game.js
function pick (array) {
  return array[array.length * Math.random() | 0]
}

function pickremove(array){
  var i = array.length * Math.random() | 0;
  var el = array[i]
  array.splice(i, 1)
  return el
}

function nop () {}

function Card (props) {
  this.image = props.image || ''
  this.title = props.title
  this.describe = props.describe
  this.options = props.options
  this.environment = props.environment
  this.applicable = props.applicable || function () { return true; }
}

Card.prototype = {
  clone: function () {
    return new Card(this)
  }
}

function Player () {
  this.buff = {}
}

function Effect (props) {
  this.option = props.option
  this.resolve = props.resolve || nop
}
Effect.prototype = {
  clone: function () { return new Effect(this) }
}

function Game (random) {
  this.random = random
  this.resolved = []
  this.deck = []
  this.player = new Player()
}

Game.prototype = {
  reset: function () {
    this.player = new Player()
    this.resolved = []
    this.deck = []
    
    var partial = [];
    for(var i = 0; i < 30; i++){
      if(partial.length == 0){
        partial = this.random.slice()
      }
      this.deck.push(pickremove(partial))
    }

    this.deck[0] = Journey
    this.deck[15] = Aging
    this.deck[29] = DeathByAging
  },
  get activeCard() {
    return this.deck[0]
  },
  get desc() {
    return this.activeCard.describe(this)
  },
  get options() {
    return this.activeCard.options(this)
  },
  get lastResolution() {
    return this.resolved[this.resolved.length - 1]
  },
  insertAt: function (n, card) {
    if (n >= this.deck.length) {
      this.deck.push(card)
    } else {
      this.deck.splice(n, 0, card)
    }
  },
  select: function (option) {
    var card = this.deck.shift()
    var options = card.options(this)
    var effect = options[option]
    var desc = effect.resolve(this, card, effect)

    this.resolved.push({
      card: card,
      desc: desc,
      selected: option,
      effect: effect
    })

    this.dropDuplicates()

    return this.lastResolution
  },
  dropDuplicates: function(){
    var avoid = this.resolved.slice(this.resolved.length-3)
    while(this.deck.length > 0){
      var card = this.deck[0]
      
      // drop cards that are not applicable
      if(!card.applicable(this)){
        this.deck.shift()
        continue
      }
      
      // drop cards that have recently been drawn
      var skip = false
      for(var k = 0; k < avoid.length; k++){
        if(avoid[k].card.title === card.title){
          skip = true
        }
      }
      if(!skip){
        break
      }
      this.deck.shift()
    }
  },
  yes: function () { return this.select('yes'); },
  no: function () { return this.select('no'); }
}

File: Assets/World/Game.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

namespace Game
{
	public abstract class Card
	{
		public string title = "";
		public string image = "";
		public string environment = "";
		public int encounters = 0;

		// applicable checks whether the card should be shown
		public abstract bool applicable (State state);

		// describes the current situation for this card
		public abstract string[] describe (State state);

		// lists options available here
		public abstract Options options (State state);

		// Option describes a single choice that a person can make on a card
		public struct Option
		{
			// title of the option
			public string title;
			// callback to resolve the effect
			public OnResolve resolve;

			public delegate string[] OnResolve ();
		}

		// options is the two choices a player can make in a particular situation
		public struct Options
		{
			public Option yes;
			public Option no;
		}

		// class for tracking choices that the player made throughout the game
		public class Choice
		{
			public Card card;
			public Option option;
			public string selected;
			public string description;
		}
	}

	// Player lists all the possible buffs
	public class Player
	{
		public bool hut = false;
		public bool flu = false;
		public bool stickyBoots = false;
		public bool creepingTerror = false;
		public bool depression = false;
		public bool map = false;
		public bool corpsePoker = false;

		public bool oldAge = false;

		public bool jasmine = false;
		public bool dianne = false;
		public bool noemi = false;
		public bool seenDeath = false;

		public bool ghostlyLady = false;
		public bool deliriousVisions = false;
		public bool peaceOfMind = false;
	}

	// State manages and updates the game state
	public class State
	{
		public Player player = new Player ();
		public List<Card> deck = new List<Card> ();
		public List<Card.Choice> history = new List<Card.Choice> ();
		public World world = new World ();

		public Card currentCard = null;
		public string currentDescription = "";
		public Card.Options currentOptions = new Card.Options ();
		public Card.Choice lastChoice = null;

		public void setup ()
		{
			deck.Clear ();
			history.Clear ();

			world = new World ();

			List<Card> encounters = world.AllEncounters ();
			for (int i = 0; i < 30; i++) {
				if (encounters.Count == 0) {
					encounters = world.AllEncounters ();
				}
				deck.Add (Rand.Pick (encounters));
			}

			deck [deck.Count - 5] = world.Ageing;
			deck [deck.Count - 1] = world.LiteralDeath;

			updateCurrentCard (world.Journey);
		}

		public bool deckEmpty ()
		{
			return deck.Count == 0;
		}

		private void updateCurrentCard (Card card)
		{
			currentCard = card;
			if (card == null) {
				currentDescription = "";
				currentOptions = new Card.Options ();
				return;
			}
			currentDescription = Text.Sanitize (card.describe (this));
			currentOptions = card.options (this);
		}

		private Card.Choice choose (string selected, Card.Option option)
		{
			currentCard.encounters++;

			Card.Choice choice = new Card.Choice ();
			choice.card = currentCard;
			choice.description = Text.Sanitize (option.resolve ());
			choice.option = option;
			choice.selected = selected;

			history.Add (choice);

			avoidDuplicates ();

			// advance to the next card
			lastChoice = choice;
			if (deck.Count == 0) {
				updateCurrentCard (null);
			} else {
				Card card = deck [0];
				deck.RemoveAt (0);
				updateCurrentCard (card);
			}

			return choice;
		}

		private List<Card.Choice> lastChoices (int n)
		{
			int low = history.Count - 4;
			int high = low + 4;
			if (low < 0) {
				low = 0;
			}
			if (high >= history.Count) {
				high = history.Count;
			}
			return history.GetRange (low, high - low);
		}

		private void avoidDuplicates ()
		{
			if (history.Count == 0) {
				return;
			}

			List<Card.Choice> avoid = lastChoices (5);

			while (deck.Count > 0) {
				Card candidate = deck [0];

				// remove cards that shouldn't be shown more
				if (!candidate.applicable (this)) {
					deck.RemoveAt (0);
					continue;
				}

				// avoid repeating same card that has been shown recently
				bool skip = false;
				for (int k = 0; k < avoid.Count; k++) {
					if (avoid [k].card.title == candidate.title) {
						skip = true;
						break;
					}
				}

				if (skip) {
					deck.RemoveAt (0);
					continue;
				}
				break;
			}
		}

		public void die ()
		{
			deck.Clear ();
		}

		public Card.Choice yes ()
		{
			return choose ("yes", currentCard.options (this).yes);
		}

		public Card.Choice no ()
		{
			return choose ("no", currentCard.options (this).no);
		}
	}
}
